File: jupyter2/ps/iteration/minimize_polynomial/jl/minimize_polynomial.jl
### begin my answer

### begin hidden
function min_value(a, b, n)
    f(x) = x^4 - 3x^3 + 2x^2 + x + 1
    best = f(a)
    for i in 1:n
        x = a + (b - a) * i / n
        v = f(x)
        if v < best
            best = v
        end
    end
    best
end
### end hidden
### end my answer

function main()
    @assert abs(min_value(-1, 1, 1000) - 0.903266) < 1.0e-6
    @assert abs(min_value( 1, 2, 1000) - 1.928766) < 1.0e-6
    println("OK")
end

main()

File: jupyter2/ps/iteration/minimize_polynomial/rs/minimize_polynomial.rs
/** begin my answer */

/** begin hidden */
fn min_value(a: f64, b: f64, n: i64) -> f64 {
    let f = |x: f64| x*x*x*x - 3.0*x*x*x + 2.0*x*x + x + 1.0;
    let mut best = f(a);
    for i in 1..=n {
        let x = a + (b - a) * (i as f64) / (n as f64);
        let v = f(x);
        if v < best {
            best = v;
        }
    }
    best
}
/** end hidden */
/** end my answer */

fn main() {
    assert!((min_value(-1.0, 1.0, 1000) - 0.903266).abs() < 1.0e-6);
    assert!((min_value( 1.0, 2.0, 1000) - 1.928766).abs() < 1.0e-6);
    println!("OK");
}
